docs/vhlo.md

```diff
@@ -133,33 +133,28 @@ negative unit tests of the feature, as well as compatibility unit tests.
+before version `X`. If adding a new version of an op, add a test file like
+`vhlo_to_version_downgrade.0_Y_0.mlir` that shows that the op can be downgraded
+successfully.
 
-A few examples:
-
-* Backward compatibility, positive tests: [vhlo_to_version_upgrade.mlir](https://github.com/openxla/stablehlo/blob/6886b59f6cd4369674e7e3beff61301c145176e2/stablehlo/tests/vhlo_to_version_upgrade.mlir#L2)
-* Forward compatibility, positive tests: [vhlo_to_version_downgrade.mlir](https://github.com/openxla/stablehlo/blob/6886b59f6cd4369674e7e3beff61301c145176e2/stablehlo/tests/vhlo_to_version_downgrade.mlir#L1)
-* Forward compatibility, negative tests: [vhlo_to_version_downgrade_invalid.0_9_0.mlir](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/vhlo/vhlo_to_version_downgrade_invalid.0_9_0.mlir)
+If your op has default attributes, include tests that show that the defaults are
+serialized and deserialized correctly.
 
 ### Add Versioned Serialization Test
 
-After adding a test point to `stablehlo_legalize_to_vhlo.mlir`, create a
-versioned copy of the file named `stablehlo_legalize_to_vhlo.0_X_0.mlir` as
-follows, along with a bytecode version of said file with a `.0_X_0.mlir.bc`
-extension. Add [proper FileCheck lines](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.0_9_0.mlir#L1-L3)
-for forward and backward compatibility testing.
+After adding tests to `stablehlo_legalize_to_vhlo.mlir`, copy the versioned test
+file with the largest version into a new file at the new version, and add the
+new tests to that file as well. You will need to also create an associated
+bytecode file using `stablehlo-translate`:
 
 ```bash
-$ export TARGET_VERSION=0.X.0
-$ export TARGET_FILENAME=${TARGET_VERSION//./_}
-$ cp stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.mlir stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.$TARGET_FILENAME.mlir
-$ build/bin/stablehlo-translate --serialize --target=$TARGET_VERSION --strip-debuginfo stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.$TARGET_FILENAME.mlir > stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.$TARGET_FILENAME.mlir.bc
-
-# Replace RUN commands in stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.0_X_0.mlir with the following for 0.X.0:
-// RUN: stablehlo-opt --mlir-print-op-generic %s.bc | FileCheck %s
-// RUN: stablehlo-translate --deserialize %s.bc | stablehlo-translate --serialize --target=0.X.0 | stablehlo-opt --mlir-print-op-generic | FileCheck %s
-// RUN: diff <(stablehlo-translate --deserialize %s.bc | stablehlo-opt) <(stablehlo-opt --strip-debuginfo %s)
-// RUN: diff %s.bc <(stablehlo-translate --serialize --target=0.X.0 --strip-debuginfo %s)
+export TARGET_VERSION=0.X.0
+export TARGET_FILENAME=${TARGET_VERSION//./_}
+stablehlo-translate --serialize --target=$TARGET_VERSION --strip-debuginfo stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.$TARGET_FILENAME.mlir > stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.$TARGET_FILENAME.mlir.bc
```

Comment:
I can see you have a prose at the top, but the bash instruction are useful as well (like `cp` instruction) if we are specifying the whole pipeline.

Reply:
IMO `cp` is simple enough to not warrant a bash example. My fear is that if we have a script that "seems" complete, developers may apply the script blindly and miss some steps (like I did :upside_down_face:). To be honest I would personally be in favor of removing any bash examples, as it is easy to figure out how to generate the bytecode by looking at the tests. WDYT?

Reply:
I am on board with that.